### DISCUSSION TITLE
@ngrx/data doesn't seem to tolerate parameters

### DISCUSSION THREAD

**Discussion by wbhob** [ngrx/platform#3093] [Ideas]
## Other information:
_Many_ apps have relational data, as well as child structures, and I'm surprised an issue hasn't been opened about this before. Essentially, our use case is as follows.

We have a base object, say "vendors", which have products and events. Managing the vendors with @ngrx/data is relatively simple, but as soon as I try to write a data wrapper for events or products, type-checking prevents me from doing so.

In other words, vendors can be fetched from `/api/vendors`, but getting vendor products requires passing a vendor id, i.e. `/api/vendors/:vendor_id/products`. There aren't any docs on this whatsoever, and, like I said, the built in types don't like this fairly common data structure _at all_.

Similarly, there is no documentation on how to add peripheral data, like pagination. Other issues about this have said that ngrx/data doesn't care about your pagination, and that you can implement it yourself, which would be fine if there were some sort of sample app or demo or docs to base it on. But after hunting the internet, there don't seem to be any docs or demos on a *sample* implementation that we can base it on, just people saying "ngrx data lets you do it yourself!".

## I would be willing to submit a PR for the docs :heart:

[X] Yes (Assistance is provided if you need help submitting a pull request)
[ ] No

**Comment by wbhob** [comment]
For clarity:

All we need is to pass a parameter in to, say, `getAll`. In my container code, I am already aware of the `vendor_id`, and the only thing stopping me is that I can't pass a parameter to `getAll()` or any other method. It would be life saving to allow for parameterized methods in the Collection and Data service models, and to pass them accordingly under the hood.

**Comment by JiaLiPassion** [comment]
@wbhob, I have the same use cases with you, and I am working on the documentation and PR of those issues, for the `pagination`, you can reference to my updated doc here https://github.com/ngrx/platform/blob/master/projects/ngrx.io/content/guide/data/entity-metadata.md

**Comment by timdeschryver** [comment]
@JiaLiPassion, this is #1944 right?

**Comment by JiaLiPassion** [comment]
@timdeschryver. #1944 is another use case. @wbhob 's use cases are.
1. pagination, this can be resolved by my earlier PR, #1921.
2. nested data, such as products belongs to vendor, so get products need to pass vendor id, this one is not supported by `@ngrx/data` now.

#1944 resolved another use case, which is we may have multiple backend endpoints return the similar data, such as `getMostPopularProducts` and `getLatestProducts`, if we want to reuse the `getWithQuery action`, we need to use `tag`. (the `tag` by design is to resolve the use cases like this).

**Comment by timdeschryver** [comment]
Ok, thanks for clarifying this.

**Comment by wbhob** [comment]
Thanks for the help guys! I'll keep an eye out for this issue, as this would help a lot with our development.

**Comment by GreedyA1** [comment]
@JiaLiPassion Is there any workaround or updates on nested data?

Could you recommend how to deal with sitaution, where we have nested data from server?

**Comment by AdditionAddict** [comment]
There's a hacky way using `EntityCacheDispatcher`

- Create your custom action e.g. type "[OrderProductsPage ONINIT] ONINIT"
- Create effect that calls saveEntities
```
// include in custom action payload and get this id here
const id;

// fake changes to get past guards in EntityCacheEffects
const changeSet = {
    changes: [
        cif.add("Product", {}) // required to get past guard
    ],
    // hook for defining custom behaviour
    tag: "CUSTOM_GET_PRODUCTS" 
};
this.entityCacheDispatcher.saveEntities(changeSet, `orders/${id}/products`)
```
-  a. Extend `EntityCacheDataService` and override the `saveEntities` method to first check if tag is CUSTOM_GET_PRODUCTS otherwise call `super.saveEntities(changeSet, url)` 
- b. If tag is CUSTOM_GET_PRODUCTS than call get http and pipe map to changeSet with ChangeSetOperation.Add
- pipe map to `ChangeSet` so that SAVE_ENTITIES_SUCCESS reducer kicks in

### [Stackblitz](https://stackblitz.com/edit/angular-rkok45-custom-url-ngrx-data)
---

This begs the natural question...

@wardbell should `EntityCacheDispatcher` be extended to include method with following API?

`getEntities(url: string, params?: QueryParams, mergeStrategy?: MergeStrategy): Observable<ChangeSet>`

This way someone can call a get request with whatever url they like and change state in a manner of their choosing by mapping to an `Observable<ChangeSet>` response and specifying `mergeStrategy`

Would require following changes:

_New Entity Cache Actions_
GET_ENTITIES
GET_ENTITIES_SUCCESS
GET_ENTITIES_FAIL

Implementations in _EntityCacheEffects_ _EntityCacheDataService_. Can't find anything else...

Happy to do a PR

As an aside, do you have a UML diagram / fuller architecture diagram for ngrx/data?

**Comment by brandonroberts** [comment]
If someone still wants to open a PR for this, feel free

**Comment by krishnanmk** [comment]
@brandonroberts Would love to contribute. How can I get started? Does the following make sense?

My initial thought is to simply support Angular HTTP Client's new ["context" token for HTTP interceptor](https://netbasal.com/308a1ca2f3dd) feature when calling ANY of the side-effect actions in EntityCollectionService, i.e. getAll, getWithQuery, getByKey, add, update, delete, upsert.
That way, the core ngrx data need not be changed at all. People who want different behaviours for their http calls can introduce the "context" and pass it as an EntityActionOptions param AND use HttpInterceptor to modify the final HTTP API call signature/behaviour as they like.

I feel that this approach is highly scalable because the responsibility of constructing the API URL is delegated to the user (developer) himself.

Shall I get started on this approach?

**Reply by smorandi** [reply]
i still would love to be able to access the „context“ information in my customized data-service. i don‘t get why it would be such an issue to pass along the action itself, or for that matter the entity-action-option into the data-service itself.
i would rather love to do specific thing like manipulating backend-calls where it actually happens, i.e. in the data-service itself, and not in a custom http-interceptor which contains logic i. a different place.

while i see the point of having a centralized place to process things, i would like to have everything i need in order to do backend-calls all in the data-service without having to distribute processing logic in different areas

**Reply by cam-m** [reply]
As we discussed in #3663 seems PR #3664 will give you a reasonable workaround by allowing you to pass data (query parameters) in EntityActionOptions into EntityServerCommands and have that data available in your custom Data Service.

**Comment by cam-m** [comment]
Hi @brandonroberts et al, I found I had some similar requirements, at least for the pagination aspects. Have created a pull request #3664 which adds support for passing http query parameters and/or http headers to api requests via angular's `HttpParamsOptions`. Hope its useful.
